File: hu/codes/python/chapter_computational_complexity/space_complexity.py
"""
File: space_complexity.py
Created Time: 2022-11-25
Author: krahets (krahets@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import ListNode, TreeNode, print_tree


def function() -> int:
    """Függvény"""
    # Néhány műveletet végez
    return 0


def constant(n: int):
    """Konstans rend"""
    # Konstansok, változók, objektumok O(1) tárhelyet foglalnak
    a = 0
    nums = [0] * 10000
    node = ListNode(0)
    # A ciklus változói O(1) tárhelyet foglalnak
    for _ in range(n):
        c = 0
    # A ciklus függvényei O(1) tárhelyet foglalnak
    for _ in range(n):
        function()


def linear(n: int):
    """Lineáris rend"""
    # Egy n hosszúságú lista O(n) tárhelyet foglal
    nums = [0] * n
    # Egy n méretű hash tábla O(n) tárhelyet foglal
    hmap = dict[int, str]()
    for i in range(n):
        hmap[i] = str(i)


def linear_recur(n: int):
    """Lineáris rend (rekurzív megvalósítás)"""
    print("Rekurzió n =", n)
    if n == 1:
        return
    linear_recur(n - 1)


def quadratic(n: int):
    """Négyzetes rend"""
    # Egy 2D lista O(n^2) tárhelyet foglal
    num_matrix = [[0] * n for _ in range(n)]


def quadratic_recur(n: int) -> int:
    """Négyzetes rend (rekurzív megvalósítás)"""
    if n <= 0:
        return 0
    # A nums tömb hossza: n, n-1, ..., 2, 1
    nums = [0] * n
    return quadratic_recur(n - 1)


def build_tree(n: int) -> TreeNode | None:
    """Exponenciális rend (teljes bináris fa felépítése)"""
    if n == 0:
        return None
    root = TreeNode(0)
    root.left = build_tree(n - 1)
    root.right = build_tree(n - 1)
    return root


"""Fő kód"""
if __name__ == "__main__":
    n = 5
    # Konstans rend
    constant(n)
    # Lineáris rend
    linear(n)
    linear_recur(n)
    # Négyzetes rend
    quadratic(n)
    quadratic_recur(n)
    # Exponenciális rend
    root = build_tree(n)
    print_tree(root)

File: hu/codes/java/chapter_computational_complexity/space_complexity.java
/**
 * File: space_complexity.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_computational_complexity;

import utils.*;
import java.util.*;

public class space_complexity {
    /* Függvény */
    static int function() {
        // Néhány műveletet végez
        return 0;
    }

    /* Konstans rend */
    static void constant(int n) {
        // Konstansok, változók, objektumok O(1) tárhelyet foglalnak
        final int a = 0;
        int b = 0;
        int[] nums = new int[10000];
        ListNode node = new ListNode(0);
        // A cikluson belüli változók O(1) tárhelyet foglalnak
        for (int i = 0; i < n; i++) {
            int c = 0;
        }
        // A cikluson belüli függvények O(1) tárhelyet foglalnak
        for (int i = 0; i < n; i++) {
            function();
        }
    }

    /* Lineáris rend */
    static void linear(int n) {
        // n hosszúságú tömb O(n) tárhelyet használ
        int[] nums = new int[n];
        // n hosszúságú lista O(n) tárhelyet foglal
        List<ListNode> nodes = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            nodes.add(new ListNode(i));
        }
        // n hosszúságú hash tábla O(n) tárhelyet foglal
        Map<Integer, String> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(i, String.valueOf(i));
        }
    }

    /* Lineáris rend (rekurzív implementáció) */
    static void linearRecur(int n) {
        System.out.println("Recursion n = " + n);
        if (n == 1)
            return;
        linearRecur(n - 1);
    }

    /* Exponenciális rend */
    static void quadratic(int n) {
        // Mátrix O(n^2) tárhelyet használ
        int[][] numMatrix = new int[n][n];
        // 2D lista O(n^2) tárhelyet használ
        List<List<Integer>> numList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            List<Integer> tmp = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                tmp.add(0);
            }
            numList.add(tmp);
        }
    }

    /* Négyzetes rend (rekurzív implementáció) */
    static int quadraticRecur(int n) {
        if (n <= 0)
            return 0;
        // A nums tömb hossza n, n-1, ..., 2, 1
        int[] nums = new int[n];
        System.out.println("In recursion n = " + n + ", nums length = " + nums.length);
        return quadraticRecur(n - 1);
    }

    /* Főprogram */
    static TreeNode buildTree(int n) {
        if (n == 0)
            return null;
        TreeNode root = new TreeNode(0);
        root.left = buildTree(n - 1);
        root.right = buildTree(n - 1);
        return root;
    }

    /* Főprogram */
    public static void main(String[] args) {
        int n = 5;
        // Konstans rend
        constant(n);
        // Lineáris rend
        linear(n);
        linearRecur(n);
        // Exponenciális rend
        quadratic(n);
        quadraticRecur(n);
        // Exponenciális rend
        TreeNode root = buildTree(n);
        PrintUtil.printTree(root);
    }
}
